### ISSUE TITLE
Online dependencies

### ISSUE THREAD

**Issue by Nixellion** [danielperna84/hass-configurator#177]
Yesterday had internet outage for a couple hours and wanted to create a binary sensor to ping google and notify me when internet gets back. Unforunately I could not use Configurator because it was trying to load google fonts and something else, and it took forever to load.

TLDR; It would be great if configurator did not have any online dependencies like fonts.

**Comment by danielperna84** [comment]
This is related to issue #145 and #73 and has been discussed there already. The readme also states this early at the top:

> IMPORTANT: The configurator fetches JavaScript libraries, CSS and fonts from CDNs. Hence it does NOT work when your client device is offline. And it is only available for Python 3.

It is more than just fonts. [Here](https://github.com/danielperna84/hass-configurator/tree/offline) is a branch where I started implementing this by including most of the libraries. But with the AceExplorer it's a ton of files that would have to be included, and in the current state the included webserver isn't capatble of delivering the files at the structure the AceEditor expects them. So even though I could get part of this offline, the main library (the editor) is a problem that causes a lot of work.
Whoever feels the need to implement his can freely do so and create a PR once done. As mentioned before, I've done it already for CSS etc., so there's a rough guideline how it could be done.

**Comment by danielperna84** [comment]
I now have applied the changes from the branch mentioned above into the current devel-branch. So at least some files will by served locally due to the change. Not included is the Ace Editor and the CSS/fonts of materialize. Serving those with my primitive implementation wont's work without modifying the libraries, and for that I'm too lazy. But with my latest commit it should be easier for someone else to take care of that.

**Comment by Misiu** [comment]
@danielperna84 I think than moving to offline support is a good decision.
I just wanted to ask why you use Ace Editor? Home Assistant is using Codemirror.

**Comment by danielperna84** [comment]
@Misiu 
Because I have created the configurator long before Home Assistant had text editing capability, and back then Ace Editor seemed to be a good choice. It was easy to implement and did it's job.

**Comment by fake-name** [comment]
I just hit this issue too. I have 3rd party scripts blocked by default, and the editor both fails to work and has no error feedback.

The entire point of homeassistant is to *not* rely on cloud anything. I was **extremely** surprised anything depends on files served from anywhere on the internet.

**Comment by pubdc** [comment]
Indeed : the file-editor wont work unless it can access CloudFlare. One of the main reasons I opted for HA is that everything is local and secure and won't halt if the house goes offline. So I am disappointed about the need to access CloudFlare. Furthermore it is not an elegant failure : it just refuses to access/open the file, there is no error message that states "hey I cant access cloudflare"

**Comment by starsoccer** [comment]
+1 to the above. HA keeps pushing it self as the no cloud solution yet when it actually comes to there being no cloud it fails and features that dont really need cloud outright break

**Comment by demlak** [comment]
any news on this? whatabout caching those files localy?

I am not a coder.. what about `ace.config.set("basePath", "https://url.to.a/folder/that/contains-ace-modes"); --> https://ace.c9.io/#about=&nav=howto
`

**Comment by JanMR21** [comment]
This happens to me too, I am running the adguard add on which filters out all cloudflare resources... because - well - there is this privacy issue...

**Comment by OdinVex** [comment]
> any news on this? whatabout caching those files localy?
> 
> I am not a coder.. what about `ace.config.set("basePath", "https://url.to.a/folder/that/contains-ace-modes"); --> https://ace.c9.io/#about=&nav=howto `

They can be used locally with relative URIs. It makes zero sense for this project to use online-only resources unless their aim were malicious, frankly.

**Comment by danielperna84** [comment]
> They can be used locally with relative URIs. It makes zero sense for this project to use online-only resources unless their aim were malicious, frankly.

Actually, this project dates back to when HAOS didn't even exist, and everyone had to do manual core-installations. For that reason I chose to include everything within one single Python file, which was easy to "install" for novice users. Because of that, the included "webserver" isn't really a webserver, but instead has hardcoded paths, which deliver only a limited set of resources. So no routing whatsoever. Therefore it's not just as easy as using relative paths, because the backend has to be completely rewritten to support nested paths, while not breaking the existing functionality. For some stuff it was easy to add the paths quickly. But the Ace editor has a gazillion files, nested within sub-folders.

Feel free to contribute. I have become a parent and bought a house, both consuming all of my time until further notice.

**Comment by OdinVex** [comment]
The web browser can use relative paths to access static local resources. It's about packaging and use. I've found too many issues with HAOS and "privacy/it's really just upsell-ware", I've only been following up to note.

Congratulations on the life-happenings. :]

**Comment by guytas** [comment]
OK, any alternative solution to the file editor?  There is no way for me to use internet every time I need to modify a file!

**Comment by jf-uu** [comment]
Studio Code Server seems to be fully local from what I can tell.

**Comment by Misiu** [comment]
https://community.home-assistant.io/t/home-assistant-community-add-on-visual-studio-code/107863
I was going to suggest the same addon. I'm using it for more than a month, works perfect.

**Comment by guytas** [comment]
Thanks!

**Comment by daniporr** [comment]
> Studio Code Server seems to be fully local from what I can tell.

Unfortunately this is not entirely true.
The container connects to Microsoft services and also includes an additional repository.

Moreover, in most cases a simple editor is all that is needed, rather than the advanced features of VS Code. And VS Code is causing considerable problems for HA users (see https://community.home-assistant.io/t/home-assistant-community-add-on-visual-studio-code/107863/641).

We really need to ensure it works fully locally!

**Comment by daniporr** [comment]
I had a look at the "offline" branch Daniel posted in https://github.com/danielperna84/hass-configurator/issues/177#issuecomment-567966058
I am certainly no expert, but having checked the changes to the HTML, CSS and JS files, they all appear reasonable. But I have no idea regarding the changes in the Python file, as I don't understand their meaning.
@danielperna84 if you have time, could you please provide some guidance on what is missing or what needs to be improved in that branch to make it stable enough to be merged into master and included in the next release?

I'd also like to add that even a small improvement is better than none. For instance, we could start by removing the dependency on Google services for the fonts and icons, and then localise jQuery and js-yaml, and so on...

**Comment by danielperna84** [comment]
@daniporr 

That was 5 years ago. I don't recall the details. From what I've written back then I guess I just added some of the files statically, along with the required endpoints.

That being said, I don't have the capacity to further work on this project. Home Assistant has thousands of contributors, who should be capable to either continue my work, or rebuild it way more efficiently. As long as they don't do so, nothing will happen.

**Comment by daniporr** [comment]
Thank you for your honest feedback on this!

Hopefully someone will be able to help on this and finalise the code in that branch.

**Comment by guytas** [comment]
> Studio Code Server seems to be fully local from what I can tell.

Ok, I used studio code server for a while before I realised that it was slowing down the system like crazy. It is using 98% of the cpu power and keep incrementing its memory usage untill it runs out. After I uninstalled it, the cou power went down to 1 to 2%. 

Not sure I understand all this thread, but I wish I could download the CSS and other files missing and set the file editor to use the local files instead of pulling them from the net.  Is it possible? (Sorry if this was already explained but sometime I miss some tips in english)

**Comment by daniporr** [comment]
> Not sure I understand all this thread, but I wish I could download the CSS and other files missing and set the file editor to use the local files instead of pulling them from the net. Is it possible? (Sorry if this was already explained but sometime I miss some tips in english)

@guytas as mentioned above by danielperna84, they created a new branch (https://github.com/danielperna84/hass-configurator/tree/offline) where they started to work on exactly what you mentioned, but they had no time to complete all the changes needed.

If you are able to help and want to complete the work started in that branch, I believe danielperna84 will be glad to accept your contribution and we will all be happy to finally get a fully local version.
